File: Algorithms/C++/Array/189_Rotate_Array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        if (nums.empty()) {
            return;
        }

        int n = nums.size();
        k = k % n;
        if (k == 0) {
            return;
        }

        reverse(nums, 0, n - k - 1);
        reverse(nums, n - k, n - 1);
        reverse(nums, 0, n - 1);

        return;
    }
private:
    void reverse(vector<int>& nums, int start, int end) {
        while (start < end) {
            int tmp = nums[start];
            nums[start] = nums[end];
            nums[end] = tmp;

            start++;
            end--;
        }
    }
};

File: Algorithms/Swift/Easy/189_Rotate_Array.swift
class Solution {
    func rotate(inout nums: [Int], _ k: Int) {
        if nums.isEmpty {
            return
        }

        let stepsToRotate = k % nums.count
        if stepsToRotate == 0 {
            return
        }

        rotateArray(&nums, from: 0, to: nums.count - stepsToRotate - 1)
        rotateArray(&nums, from: nums.count - stepsToRotate, to: nums.count - 1)
        rotateArray(&nums, from: 0, to: nums.count - 1)
    }

    private func rotateArray(inout nums: [Int], from startIndex: Int, to endIndex: Int) {
        if startIndex >= endIndex {
            return
        }

        var start = startIndex, end = endIndex
        while start < end {
            let tmp = nums[start]
            nums[start] = nums[end]
            nums[end] = tmp

            start += 1
            end -= 1
        }
    }
}
